connectors/mongo/progress.go

```diff
@@ -0,0 +1,118 @@
+package mongo
+
+import (
+	"log/slog"
+	"math"
+	"fmt"
+	"github.com/adiom-data/dsync/protocol/iface"
+)
+
+// restoreProgressDetails restores the progress metrics from the persisted tasks and progress
+func RestoreProgressDetails(status *iface.ConnectorStatus, tasks []iface.ReadPlanTask) { 
+	slog.Debug("Restoring progress metrics from tasks")
+	status.ProgressMetrics.TasksTotal = int64(len(tasks))
+	for _, task := range tasks {
+		ns := iface.Namespace{Db: task.Def.Db, Col: task.Def.Col}
+		nsStatus := status.ProgressMetrics.NamespaceProgress[ns]
+		//check if the namespace status exists, if not create it
+		if nsStatus == nil {
+			nsStatus = &iface.NamespaceStatus{
+				EstimatedDocCount:   0,
+				Throughput:          0,
+				Tasks:               []iface.ReadPlanTask{},
+				TasksCompleted:      0,
+				TasksStarted:        0,
+				DocsCopied:          0,
+				EstimatedDocsCopied: 0,
+				ActiveTasksList:     make(map[iface.ReadPlanTaskID]bool),
+			}
+			status.ProgressMetrics.NamespaceProgress[ns] = nsStatus
+		}
+		nsStatus.Tasks = append(nsStatus.Tasks, task)
+		nsStatus.EstimatedDocCount += task.EstimatedDocCount
+		//if the task is completed, update the document counters
+		if task.Status == iface.ReadPlanTaskStatus_Completed {
+			status.ProgressMetrics.TasksCompleted++
+			status.ProgressMetrics.NumDocsSynced += task.DocsCopied
+
+			nsStatus.TasksCompleted++
+			nsStatus.DocsCopied += task.DocsCopied
+
+			nsStatus.EstimatedDocsCopied += task.EstimatedDocCount
+		}
+	}
+	status.ProgressMetrics.NumNamespaces = int64(len(status.ProgressMetrics.NamespaceProgress))
+
+	for ns, nsStatus := range status.ProgressMetrics.NamespaceProgress {
+		if nsStatus.TasksCompleted == int64(len(nsStatus.Tasks)) {
+			status.ProgressMetrics.NumNamespacesCompleted++
+		}
+		status.ProgressMetrics.Namespaces = append(status.ProgressMetrics.Namespaces, ns)
+	}
+
+	slog.Debug(fmt.Sprintf("Restored progress metrics: %+v", status.ProgressMetrics))
+}
+
+// update estimated namespace doc counts from the actual database
+func ResetNsProgressEstimatedDocCounts(mc *BaseMongoConnector) error {
+	for ns, nsStatus := range mc.Status.ProgressMetrics.NamespaceProgress {
+		collection := mc.Client.Database(ns.Db).Collection(ns.Col)
+		count, err := collection.EstimatedDocumentCount(mc.Ctx)
+		if err != nil {
+			return fmt.Errorf("failed to count documents: %v", err)
+		}
+		nsStatus.EstimatedDocCount = int64(count)
+	}
+	return nil
+}
+
+// Updates the progress metrics once a task has been started
+func TaskStartedProgressUpdate(mc *BaseMongoConnector, nsStatus *iface.NamespaceStatus, taskId iface.ReadPlanTaskID) {
+	mc.muProgressMetrics.Lock()
```

Comment:
these helper methods are accessing private members like mc.muProgressMetrics.Lock(), which is a bit of a smell.

We can create a progress component that holds all the progress related things like locks, and instead of the connector having all of these, it will simply have that progress component.

So this function could for example be just a method of that component.